### PR TITLE
[VP-28] Extend build system for visionOS

### PR DESCRIPTION
**Pull request by ViktorHeisenberger**
## Jira ticket URL (Why?)
[VP-28](https://shapr3d.atlassian.net/browse/VP-28)

## Short description (What? How?) 📖
This PR extends Filament's build system, to allow building for visionOS. This requires adding cases to `build.sh` for visionOS, and extending the cmake scripts. Since Xcode 15.2, there's no support for Intel architectures in visionOS Simulator, so only ARM64 builds are needed.

The only required code change is modifying a build-time condition to only be true for Catalyst or macOS builds (no iOS, iOS Simulator, or visionOS).

ℹ️ This PR is safe to merge, no builds on older systems will break.

## Material shader statistics implications
N/A

## Upstreaming scope
Questionable. We don't know if the Google Filament team has plans for visionOS support.

### Design review 🎨
N/A

### Affected areas 🧭
`build.sh`, cmake scripts

### Special use-cases to test 🧷
Builds for existing platforms should be unaffected, building for visionOS is possible with Xcode 15.x

### How did you test it? 🤔
Manual 💁‍♂️
Built the library for iOS, iOS Simulator, Catalyst, visionOS. Tried it in Visualization on all currently supported platforms, it built, and nothing seemed broken.
Automated 💻
❌ 

[VP-28]: https://shapr3d.atlassian.net/browse/VP-28?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -709,6 +711,41 @@ function build_ios {
...
+function build_visionos {
+    local old_ogl_ios_option=${OPENGL_IOS_OPTION}
+    OPENGL_IOS_OPTION="-DFILAMENT_SUPPORTS_OPENGL=OFF"
+
+    build_desktop_tools_for_ios "${MOBILE_HOST_TOOLS}"
+
+    if [[ "${ISSUE_DEBUG_BUILD}" == "true" ]]; then
+        build_ios_target "Debug" "arm64" "xros"
+
+        if [[ "${IOS_BUILD_SIMULATOR}" == "true" ]]; then
```
Comment (arx97) [inline build.sh]: wrong variable used
```suggestion
        if [[ "${VISIONOS_BUILD_SIMULATOR}" == "true" ]]; then
```

File: build.sh [hunk 2]
```diff
@@ -709,6 +711,41 @@ function build_ios {
...
+            archive_ios "Debug" "xrsimulator"
+        fi
+
+        archive_ios "Debug" "xros"
+    fi
+
+    if [[ "${ISSUE_RELEASE_BUILD}" == "true" ]]; then
+        build_ios_target "Release" "arm64" "xros"
+
+        if [[ "${IOS_BUILD_SIMULATOR}" == "true" ]]; then
```
Comment (arx97) [inline build.sh]: same here
```suggestion
        if [[ "${VISIONOS_BUILD_SIMULATOR}" == "true" ]]; then
```

**Review by arx97 — CHANGES_REQUESTED** [review]

File: build.sh (hunk 2 above)
Comment (ViktorHeisenberger) [reply build.sh]: Fixed in 0aa267c

File: build.sh (hunk 1 above)
Comment (ViktorHeisenberger) [reply build.sh]: Fixed in 0aa267c

**Review by arx97 — APPROVED** [review]
I checked the code (didn't tested it) 🚀

**Review by lowdesertpunk — APPROVED** [review]

File: filament/backend/src/metal/MetalBuffer.mm [hunk 3]
```diff
@@ -68,7 +68,7 @@
     // If the GPU buffer is not private, just perform the copy
     if (mBuffer.storageMode != MTLStorageModePrivate) {
         memcpy(static_cast<uint8_t*>(mBuffer.contents), src, size);
-#if !TARGET_OS_IOS || TARGET_OS_MACCATALYST
+#if TARGET_OS_MACCATALYST || TARGET_OS_OSX
```
Comment (palver123) [inline filament/backend/src/metal/MetalBuffer.mm:71]: Doesn't `initializeSupportedGpuFamilies` need an update? When building for visionOS, is `TARGET_OS_IOS` defined (if not, `MetalContext::highestSupportedFamily::apple` will be missing some info)?

Comment (palver123) [reply filament/backend/src/metal/MetalBuffer.mm:71]: And what about `MetalDriver::readPixels()`?

Comment (palver123) [reply filament/backend/src/metal/MetalBuffer.mm:71]: Same question for `SamplerStateCreator::operator()`

Comment (ViktorHeisenberger) [reply filament/backend/src/metal/MetalBuffer.mm:71]: Will be fixed in https://github.com/shapr3d/filament/pull/146.

`initializeSupportedGpuFamilies` doesn't need updating, as `@available(..., *)` will return true for visionOS, and it runs fine. `MetalDriver::readPixels()` needs an update, it's in the PR. `SamplerStateCreator::operator()` checks for old iOS devices (and the simulator) that do not support sampler comparisons, which is not necessary on visionOS.

Comment (palver123) [reply filament/backend/src/metal/MetalBuffer.mm:71]: > it's in the PR
which one?

**Pull request merged**